extractor/filesystem/language/python/uvlock/uvlock.go

```diff
@@ -0,0 +1,147 @@
+package uvlock
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/BurntSushi/toml"
+	"github.com/google/osv-scalibr/extractor"
+	"github.com/google/osv-scalibr/extractor/filesystem"
+	"github.com/google/osv-scalibr/extractor/filesystem/language/python/internal/pypipurl"
+	"github.com/google/osv-scalibr/extractor/filesystem/osv"
+	"github.com/google/osv-scalibr/plugin"
+	"github.com/google/osv-scalibr/purl"
+)
+
+type uvLockPackageSource struct {
+	Virtual  string `toml:"virtual"`
+	Git      string `toml:"git"`
+	Registry string `toml:"registry"`
+}
+
+type uvLockPackage struct {
+	Name    string              `toml:"name"`
+	Version string              `toml:"version"`
+	Source  uvLockPackageSource `toml:"source"`
+
+	// uv stores "groups" as a table under "package" after all the packages, which due
+	// to how TOML works means it ends up being a property on the last package, even
+	// through in this context it's a global property rather than being per-package
+	Groups map[string][]uvOptionalDependency `toml:"optional-dependencies"`
+}
+
+type uvOptionalDependency struct {
+	Name string `toml:"name"`
+}
+type uvLockFile struct {
+	Version  int                               `toml:"version"`
+	Packages []uvLockPackage                   `toml:"package"`
+	Groups   map[string][]uvOptionalDependency `toml:"package.optional-dependencies"`
+}
+
+// Extractor extracts python packages from uv.lock files.
+type Extractor struct{}
+
+// Name of the extractor
+func (e Extractor) Name() string { return "python/uvlock" }
+
+// Version of the extractor
+func (e Extractor) Version() int { return 0 }
+
+// Requirements of the extractor
+func (e Extractor) Requirements() *plugin.Capabilities {
+	return &plugin.Capabilities{}
+}
+
+// FileRequired returns true if the specified file matches uv lockfile patterns
+func (e Extractor) FileRequired(api filesystem.FileAPI) bool {
+	return filepath.Base(api.Path()) == "uv.lock"
+}
+
+// Extract extracts packages from uv.lock files passed through the scan input.
+func (e Extractor) Extract(ctx context.Context, input *filesystem.ScanInput) ([]*extractor.Inventory, error) {
+	var parsedLockfile *uvLockFile
+
+	_, err := toml.NewDecoder(input.Reader).Decode(&parsedLockfile)
+
+	if err != nil {
+		return []*extractor.Inventory{}, fmt.Errorf("could not extract from %s: %w", input.Path, err)
+	}
+
+	packages := make([]*extractor.Inventory, 0, len(parsedLockfile.Packages))
+
+	var groups map[string][]uvOptionalDependency
+
+	// uv stores "groups" as a table under "package" after all the packages, which due
+	// to how TOML works means it ends up being a property on the last package, even
+	// through in this context it's a global property rather than being per-package
+	if len(parsedLockfile.Packages) > 0 {
+		groups = parsedLockfile.Packages[len(parsedLockfile.Packages)-1].Groups
+	}
+
+	for _, lockPackage := range parsedLockfile.Packages {
+		if lockPackage.Source.Virtual == "." {
```

Comment:
Add a comment on why these are skipped (I assume it's because this is only set for the main "uv-lockfiles" package?)

Reply:
yeah though I'll note that's an assumption on my part technically - I tried confirming via the docs but I'm not a Python or uv expert so I might not have been using the right terms; ultimately, I figure if this turns out to be wrong in some way it'll be something we can fix, and we'll learn how we're wrong in the process 🤷

I'll add a comment capturing the idea behind this skip